File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
class Solution {
    public int[] solution(int n, String[] words) {
        int[] answer = {0, 0};
        int cnt = 0;

        outer: for(int i = 1; i < words.length; i++) {
            // 1. 전 단어의 마지막 글자와 비교하기
            char last = words[i-1].charAt(words[i-1].length()-1);
            char first = words[i].charAt(0);
            // System.out.println("last = " + last + " first = " + first);
            
            if(last != first) {
                cnt = i+1;
                // System.out.println("글자 다름");
                break outer;
            }
            
            for(int j = 0; j < i; j++) {
                if(words[i].equals(words[j])) {
                    cnt = i+1;
                    // System.out.println("단어 중복");
                    break outer;
                }
            }
        
        }
        // 3. 하나라도 충족이 안되면 해당 단어가 몇 번째인지 가져오기
        // 4. n으로 나누면 몇 번째 유저의 몇 번째 순서인지 나올 것
        int a = cnt / n;
        int b = cnt % n;
        // System.out.println("a = " + a + " b = " + b + " cnt = " + cnt);
        if(cnt == 0) return answer;
        
        
        if(b > 0) {
            a += 1;
            
            answer[0] = b;
            answer[1] = a;
        } else {
            answer[0] = n;
            answer[1] = a;
        }
        
        // System.out.println("a = " + a + " b = " + b);
        

        return answer;
    }
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>
#include <map>

using namespace std;

    
// 중복값 체크할 map 생성
map<string, int> checkMap;

bool checkWord(string s1, string s2) {
    return s1[s1.length() - 1] == s2[0];
}

bool duple(string s) {
    if(checkMap.count(s) > 0) return true;
    
    ++checkMap[s];
    return false;
}

vector<int> solution(int n, vector<string> words) {
    vector<int> answer = {0, 0};
    queue<int> q;
    
    // 사용자를 q에 1부터 n까지 집어 넣음
    for(int i = 1; i <= n; i++) {
        q.push(i);
    }
    
    // 1번은 틀릴 수 없으므로
    // 맨 뒤에 입력해주자
    q.pop();
    q.push(1);
    duple(words[0]);
    
    // words의 길이만큼 순회하며 진행
    // 단 이전 글자와 비교하기 위해 처음은 생략함(index error 배제)
    for(int i = 1; i < words.size(); i++) {
        string before = words[i-1];
        string now = words[i];
        // 지금 순서
        int peopleNum = q.front();
        // 이전 단어 맨 뒷 글자와 현재 단어 맨 앞 글자가 다르면
        // 또는 중복 단어를 말했으면
        // 끝말잇기 끝
        if(!checkWord(before, now) || duple(now)) {
            answer[0] = peopleNum;
            answer[1] = i / n + 1;
            return answer;
        }
        
        q.pop();
        q.push(peopleNum);
    }

    return answer;
}